api.lua

```diff
@@ -1,61 +1,121 @@
-			local group_identifier, group_number
-			local equals_index = string.find(target_node, "=")
-			if equals_index ~= nil then
-				group_identifier = string.sub(target_node, 7, equals_index-1)
-				-- it's possible that the string was of the form "group:blah = 1", in which case we want to trim spaces off the end of the group identifier
-				local space_index = string.find(group_identifier, " ") 
-				if space_index ~= nil then
-					group_identifier = string.sub(group_identifier, 1, space_index-1)
-				end
-				group_number = tonumber(string.sub(target_node, equals_index+1, -1))
-			else
-				group_identifier = string.sub(target_node, 7, -1)
-				group_number = "all" -- special value to indicate no number was provided
-			end
-			
+			local group_identifier, group_number = parse_group(target_node)
+
 			local group_info = hopper.groups[group_identifier]
 			if group_info == nil then
 				group_info = {}
 			end
 			if group_info[group_number] == nil then
 				group_info[group_number] = {}
 			end
-			group_info[group_number][entry[1]] = entry[3]
+			if group_info[group_number][entry[1]] == nil then
+				group_info[group_number][entry[1]] = {}
+			end
+			if group_info[group_number]["extra"] == nil then
+				group_info[group_number]["extra"] = {}
+			end
+			group_info[group_number][entry[1]]["inventory_name"] = entry[3]
+			if entry["get_inventory"] then
+				group_info[group_number][entry[1]]["get_inventory"] = entry["get_inventory"]
+			end
 			hopper.groups[group_identifier] = group_info
 			neighbor_node = "group:"..group_identifier
-			-- result is a table of the form groups[group_identifier][group_number][relative_position][inventory_name]
+			-- result is a table of the form:
+			-- groups[group_identifier][group_number][relative_position]{"inventory_name","get_inventory"}
+			-- and {"set_hopper_param2"}
 		else
 			local node_info = hopper.containers[target_node]
 			if node_info == nil then
 				node_info = {}
 			end
-			node_info[entry[1]] = entry[3]
+			if node_info[entry[1]] == nil then
+				node_info[entry[1]] = {}
+			end
+			if node_info["extra"] == nil then
+				node_info["extra"] = {}
+			end
+			node_info[entry[1]]["inventory_name"] = entry[3]
+			if entry["get_inventory"] then
+				node_info[entry[1]]["get_inventory"] = entry["get_inventory"]
+			end
 			hopper.containers[target_node] = node_info
 			neighbor_node = target_node
-			-- result is a table of the form containers[target_node_name][relative_position][inventory_name]
+			-- result is a table of the form:
+			-- containers[target_node_name][relative_position]{"inventory_name","get_inventory"}
 		end
-		
-		local already_in_neighbors = false
-		for _, value in pairs(hopper.neighbors) do
-			if value == neighbor_node then
-				already_in_neighbors = true
+
+		if not is_already_in_neighbors(neighbor_node) then
+			table.insert(hopper.neighbors, neighbor_node)
+		end
+	end
+end
+
+-- additional information about containers
```

Comment:
What kind of information is currently put in here?
What is it used for?
What is the expected format of `list`?

Please extend the comment to better understand what the purpose of this function is so that it's not necessary to read and try to understand all of its code.

Reply:
I wrote to api.md too